### PR TITLE
Non-maximum-suppression op

### PR DESCRIPTION
**Pull request by zhreshold**
Add a general box non-maximum suppression op

### CODE REVIEW

File: src/operator/contrib/nms.cc [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+namespace op {
+DMLC_REGISTER_PARAMETER(NMSParam);
+
+NNVM_REGISTER_OP(_contrib_nms)
+.add_alias("_contrib_non_maximum_suppression")
+.describe(R"code(Apply non-maximum suppression to input.
+
+The output will be sorted in descending order according to `score`. Boxes with
+overlaps larger than `overlap_thresh` will be removed, and the corresponding
+position will be recorded for backward propogation.
```
Comment (piiswrong) [inline src/operator/contrib/nms.cc]: What does this mean? How is gradient for removed boxes handled? What are the -1s doing?

File: src/operator/contrib/nms-inl.h [hunk 2]
```diff
@@ -0,0 +1,437 @@
...
+}  // nms_enum
+
+struct NMSParam : public dmlc::Parameter<NMSParam> {
+  float overlap_thresh;
+  int topk;
+  int coord_start;
+  int score_index;
+  int id_index;
+  bool force_suppress;
+  int in_typ;
```
Comment (piiswrong) [inline src/operator/contrib/nms-inl.h]: typ->type

File: src/operator/contrib/nms-inl.h [hunk 3]
```diff
@@ -0,0 +1,437 @@
...
+    DMLC_DECLARE_FIELD(overlap_thresh).set_default(0.5)
+    .describe("Overlapping(IoU) threshold to suppress object with smaller score.");
+    DMLC_DECLARE_FIELD(topk).set_default(-1)
+    .describe("Apply nms to topk boxes with descending scores, -1 to no restriction.");
+    DMLC_DECLARE_FIELD(coord_start).set_default(2)
+    .describe("Start index of the consecutive 4 coordinates.");
+    DMLC_DECLARE_FIELD(score_index).set_default(1)
+    .describe("Index of the scores.");
+    DMLC_DECLARE_FIELD(id_index).set_default(-1)
+    .describe("Optional, index of the class ids, -1 to disable id.");
```
Comment (piiswrong) [inline src/operator/contrib/nms-inl.h]: what's class id?

File: src/operator/contrib/nms-inl.h [hunk 4]
```diff
@@ -0,0 +1,437 @@
...
+  int in_typ;
+  int out_typ;
+  DMLC_DECLARE_PARAMETER(NMSParam) {
+    DMLC_DECLARE_FIELD(overlap_thresh).set_default(0.5)
+    .describe("Overlapping(IoU) threshold to suppress object with smaller score.");
+    DMLC_DECLARE_FIELD(topk).set_default(-1)
+    .describe("Apply nms to topk boxes with descending scores, -1 to no restriction.");
+    DMLC_DECLARE_FIELD(coord_start).set_default(2)
+    .describe("Start index of the consecutive 4 coordinates.");
+    DMLC_DECLARE_FIELD(score_index).set_default(1)
```
Comment (piiswrong) [inline src/operator/contrib/nms-inl.h]: What is index of scores?

File: src/operator/contrib/nms.cc [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+Examples::
+
+  x = [[0, 0.5, 0.1, 0.1, 0.2, 0.2], [1, 0.4, 0.1, 0.1, 0.2, 0.2],
+       [0, 0.3, 0.1, 0.1, 0.14, 0.14], [2, 0.6, 0.5, 0.5, 0.7, 0.8]]
+  nms(x, overlap_thresh=0.1, coord_start=2, score_index=1, id_index=0,
```
Comment (piiswrong) [inline src/operator/contrib/nms.cc]: Could you describe the format of inputs and outputs?

Comment (piiswrong) [inline src/operator/contrib/nms.cc]: like what does each axis represent?

File: src/operator/contrib/nms.cc [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+according to its original index in input.
+
+Input requirements:
+1. Input tensor have at least 2 dimensions, (n, k), any higher dims will be regarded
+as batch, e.g. (a, b, c, d, n, k) == (a*b*c*d, n, k)
+2. n is the number of boxes in each batch
+3. k is the width of each box item.
+
+By default, a box is [id, score, xmin, ymin, xmax, ymax, ...],
+additional elements are allowed.
```
Comment (piiswrong) [inline src/operator/contrib/nms.cc]: you mean ignored and passed through?

File: src/operator/contrib/nms.cc [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+ * \author Joshua Zhang
+ */
+#include "./nms-inl.h"
+#include "../elemwise_op_common.h"
+
+namespace mxnet {
+namespace op {
+DMLC_REGISTER_PARAMETER(NMSParam);
+
+NNVM_REGISTER_OP(_contrib_nms)
```
Comment (piiswrong) [inline src/operator/contrib/nms.cc]: we probably should call this something more specific to object detection. Generic nms usually deals with pixels.

Comment (piiswrong) [inline src/operator/contrib/nms.cc]: like box nsm or multibox nms

File: src/operator/contrib/nms.cc (hunk 6 above)
Comment (zhreshold) [inline src/operator/contrib/nms.cc]: Blocked for suppressed boxes, pass through to original position if kept.

File: src/operator/contrib/nms.cc (hunk 7 above)
Comment (zhreshold) [inline src/operator/contrib/nms.cc]: @piiswrong So um, rename it to box_nms? region_nms?

Comment (piiswrong) [inline src/operator/contrib/nms.cc]: box_nms sounds better

**Comment by piiswrong** [comment]
any updates on this?

**Comment by zhreshold** [comment]
```
cpp-package/include/mxnet-cpp/lr_scheduler.h:7:  #ifndef header guard has wrong style, please use: MXNET_CPP_LR_SCHEDULER_H_  [build/header_guard] [5]

cpp-package/include/mxnet-cpp/lr_scheduler.h:78:  #endif line should be "#endif  // MXNET_CPP_LR_SCHEDULER_H_"  [build/header_guard] [5]

Done processing cpp-package/include/mxnet-cpp/lr_scheduler.h

Category 'build' errors found: 2

Total errors found: 2
```

**Comment by zhreshold** [comment]
Will merge with other bounding box ops. Close this for now.
